agent/tests/unit/test_watermark_manager.py

```diff
@@ -0,0 +1,123 @@
+import pytest
+import pytz
+
+from unittest.mock import Mock
+from datetime import datetime, timedelta
+from agent import pipeline
+
+
+@pytest.mark.parametrize(
+    "now, watermark, offset, watermark_delay, bucket_size, uses_schema, er", [
+        (100, None, None, 10, '5m', True, False),
+        (100, None, None, 10, '5m', False, False),
+        (1200, 1200 - 450, 1000, 100, '5m', False, False),
+        (1200, 1200 - 450, 1000, 100, '5m', True, True),
+        (1200, 1200 - 300, 1000, 10, '5m', True, False),
+        (1200, None, 1200 - 300, 10, '5m', True, False),
+        (1200, None, 1200 - 310, 10, '5m', True, True),
+        (1800, None, 1800 - 1300, 300, '5m', True, True),
+        (3900, None, 3900 - 400, 50, '1h', True, True),
+        (900, 300, 900, 300, '5m', True, True),
+        (7600, 3600, 4500, 200, '1h', True, True),
+        (7600, 3600, 4500, 401, '1h', True, False),
+    ]
+)
+def test_should_send_watermark(now, watermark, offset, watermark_delay, bucket_size, uses_schema, er):
+    pipeline_ = Mock()
+    pipeline_.uses_schema = Mock(return_value=uses_schema)
+    pipeline_.watermark_delay = watermark_delay
+    pipeline_.periodic_watermark_config = {'bucket_size': bucket_size}
+    pipeline_.has_periodic_watermark_config = Mock(return_value=True)
+    pipeline.manager.is_running = Mock(return_value=True)
+
+    if watermark:
+        pipeline_.watermark = Mock()
+        pipeline_.watermark.timestamp = watermark
+        pipeline_.has_watermark = Mock(return_value=True)
+    else:
+        pipeline_.watermark = None
+        pipeline_.has_watermark = Mock(return_value=False)
+
+    if offset:
+        pipeline_.offset = Mock()
+        pipeline_.offset.timestamp = offset
+        pipeline_.has_offset = Mock(return_value=True)
+    else:
+        pipeline_.offset = None
+        pipeline_.has_offset = Mock(return_value=False)
+
+    watermark_manager = pipeline.manager.PeriodicWatermarkManager(pipeline_)
+    watermark_manager._get_local_now_timestamp = Mock(return_value=now)
+    assert watermark_manager.should_send_watermark() == er
+
+
+def test_should_not_send_watermark():
```

Comment:
why is this a separate test and not a part of the test above?

Reply:
maybe I should call it test_should_not_send_watermark_without_periodic_watermark_config? :)